File: compiler/tests-jsoo/compare_custom.js
// Test helpers reproducing a zarith_stubs_js-like custom block: small values
// are plain immediates while large ones are custom blocks whose custom
// operations only register a [compare] (used, like the native [compare_ext],
// to compare the block against an immediate).

//Provides: test_zt_value
function test_zt_value(z) {
  return typeof z === "number" ? z : z.v;
}

//Provides: test_zt_compare
//Requires: test_zt_value
function test_zt_compare(a, b) {
  var av = test_zt_value(a);
  var bv = test_zt_value(b);
  return av < bv ? -1 : av > bv ? 1 : 0;
}

//Provides: test_zt_register
//Requires: caml_custom_ops, test_zt_compare
function test_zt_register(unit) {
  caml_custom_ops._zt = { compare: test_zt_compare };
  return 0;
}

//Provides: test_zt_make
function test_zt_make(v) {
  return { caml_custom: "_zt", v: v };
}

File: compiler/tests-jsoo/compare_custom.ml
(* A custom block whose custom operations register a [compare] op (like
   zarith_stubs_js) must have that op used when the block is compared against
   an immediate: small zarith values are immediates and large ones custom
   blocks, so [compare (Z.of_int 5) big] has to inspect the values rather than
   blindly ordering the immediate before the block. *)

external register : unit -> unit = "test_zt_register"

external make : int -> Obj.t = "test_zt_make"

let%expect_test "compare immediate vs custom with a compare op" =
  register ();
  let small n : Obj.t = Obj.repr n in
  let p label f = Printf.printf "%s = %d\n" label (f ()) in
  (* immediate on the left *)
  p "cmp 5 (big -100)" (fun () -> compare (small 5) (make (-100)));
  p "cmp 5 (big 100)" (fun () -> compare (small 5) (make 100));
  (* immediate on the right *)
  p "cmp (big -100) 5" (fun () -> compare (make (-100)) (small 5));
  p "cmp (big 100) 5" (fun () -> compare (make 100) (small 5));
  (* equal values: the op returns 0 and comparison keeps going *)
  p "eq 5 (big 5)" (fun () -> if make 5 = small 5 then 1 else 0);
  [%expect
    {|
    cmp 5 (big -100) = 1
    cmp 5 (big 100) = -1
    cmp (big -100) 5 = -1
    cmp (big 100) 5 = 1
    eq 5 (big 5) = 1
    |}]
